File: unibakonto/src/main/java/sk/pluk64/unibakonto/meals/Meals.java
package sk.pluk64.unibakonto.meals;

import java.util.ArrayList;
import java.util.List;

public class Meals {
    public final List<DayMenu> menus;

    public Meals(List<DayMenu> menus) {
        this.menus = menus;
    }

    public static Builder builder() {
        return new Builder();
    }

    @Override
    public String toString() {
        return "Meals{" +
                "menus=" + menus +
                '}';
    }

    static class Builder {
        private final List<DayMenu> menus = new ArrayList<>();
        private DayMenu curDayMenu;
        private SubMenu curSubMenu;

        public void newDay(String day) {
            curDayMenu = new DayMenu(day);
            menus.add(curDayMenu);
        }

        public void newSubMenu(String text) {
            curSubMenu = new SubMenu(text);
            curDayMenu.add(curSubMenu);
        }

        public void addMeal(Meal meal) {
            curDayMenu.curSubMenu.meals.add(meal);
        }

        public Meals build() {
            return new Meals(menus);
        }
    }

    public static class DayMenu {
        public final String dayName;
        public final List<SubMenu> subMenus = new ArrayList<>();
        public SubMenu curSubMenu;

        public DayMenu(String dayName) {
            this.dayName = dayName;
        }

        public void add(SubMenu subMenu) {
            subMenus.add(subMenu);
            curSubMenu = subMenu;
        }
    }

    public static class SubMenu {
        public final String name;
        public final List<Meal> meals = new ArrayList<>();

        public SubMenu(String name) {
            this.name = name;
        }
    }

    public static class Meal {
        public final String name;
        public final String price;

        public Meal(String name, String price) {
            this.name = name;
            this.price = price;
        }

        @Override
        public String toString() {
            return "Meal{" +
                    "name='" + name + '\'' +
                    ", price='" + price + '\'' +
                    '}';
        }
    }
}

File: unibakontoapp/src/main/java/sk/pluk64/unibakontoapp/meals/Meals.kt
package sk.pluk64.unibakontoapp.meals

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import java.util.*

@Serializable
class Meals(val menus: List<DayMenu>) {

    override fun toString(): String {
        return "Meals{" +
            "menus=" + menus +
            '}'.toString()
    }

    class Builder {
        private val menus = ArrayList<DayMenu>()
        private var curDayMenu: DayMenu? = null
        private var curSubMenu: SubMenu? = null

        fun newDay(day: String) {
            menus.add(DayMenu(day))
            curDayMenu = menus.last()
        }

        fun newSubMenu(text: String) {
            val newSubMenu = SubMenu(text)
            curSubMenu = newSubMenu
            curDayMenu?.add(newSubMenu)
        }

        fun addMeal(meal: Meal) {
            curDayMenu?.curSubMenu?.meals?.add(meal)
        }

        fun build(): Meals {
            val filteredMenus = menus.asSequence()
                .onEach {
                    it.subMenus.removeAll(
                        it.subMenus.asSequence()
                            .onEach {
                                it.meals.removeAll(
                                    it.meals.asSequence()
                                        .filter { it.name.isBlank() }
                                )
                            }
                            .filter { it.meals.isEmpty() }
                    )
                }
                .filter { it.subMenus.isNotEmpty() }
                .toList()
            return Meals(filteredMenus)
        }
    }

    @Serializable
    class DayMenu(val dayName: String) {
        val subMenus: MutableList<SubMenu> = ArrayList()
        @Transient
        var curSubMenu: SubMenu? = null

        fun add(subMenu: SubMenu) {
            subMenus.add(subMenu)
            curSubMenu = subMenu
        }
    }

    @Serializable
    class SubMenu(val name: String) {
        val meals: MutableList<Meal> = ArrayList()
    }

    @Serializable
    class Meal(val name: String, val price: String) {

        override fun toString(): String {
            return "Meal{" +
                "name='" + name + '\''.toString() +
                ", price='" + price + '\''.toString() +
                '}'.toString()
        }
    }

    companion object {
        fun builder(): Builder {
            return Builder()
        }
        val EMPTY = Meals(emptyList())
    }
}
